arch/eventloop_posix_tcp.c

```diff
@@ -490,6 +490,66 @@ TCP_registerListenSocketDomainName(UA_ConnectionManager *cm, const char *hostnam
     return UA_STATUSCODE_GOOD;
 }
 
+#ifdef _WIN32
+static
+UA_RegisteredFD *
+findRegisteredFD(TCPConnectionManager *cm, uintptr_t connectionId) {
+    UA_RegisteredFD *rfd, *rfd_tmp;
+    LIST_FOREACH_SAFE(rfd, &cm->fds, es_pointers, rfd_tmp) {
+
+        if (rfd->fd == connectionId) {
+            return rfd;
+        }
+    }
+    return NULL;
+}
+
+static
+UA_DelayedCallback *createExplicitlyDelayedShutdownCallback(UA_ConnectionManager *cm, UA_RegisteredFD *rfd, UA_Callback callback) {
+    UA_DelayedCallback *explicitlyDelayedShutdownCallback = (UA_DelayedCallback*) UA_malloc(sizeof(UA_DelayedCallback));
+    if (explicitlyDelayedShutdownCallback == NULL) {
+        return NULL;
+    }
+    explicitlyDelayedShutdownCallback->callback = callback;
+    explicitlyDelayedShutdownCallback->application = cm;
+    explicitlyDelayedShutdownCallback->data = rfd;
+    return explicitlyDelayedShutdownCallback;
+}
+
+static
+void delayedShutdownCallback(void* application, void* data) {
+
+    UA_ConnectionManager *cm = (UA_ConnectionManager*) application;
+    TCPConnectionManager *tcm = (TCPConnectionManager*)cm;
+    UA_RegisteredFD* rfd = (UA_RegisteredFD *) data;
+
+    UA_ByteString response = tcm->rxBuffer;
+    int ret = UA_recv(rfd->fd, (char *)response.data, response.length, MSG_DONTWAIT | MSG_PEEK);
+    if(ret == 0 || ret == -1) {
+        cm->connectionCallback(cm, (uintptr_t)rfd->fd, &rfd->context,
+                               UA_STATUSCODE_BADCONNECTIONCLOSED,
+                               0, NULL, UA_BYTESTRING_NULL);
+        TCP_close(tcm, rfd);
```

Comment:
You can early-return here.
So you don't need the else-case.

Reply:
I would prefer the if/else way because it carries the semantic of what is to be done there:

"if we receive nothing or if there is a receive-error, then close the socket. Else recreate the delayed shutdown callback and add it to the loop again."

Reply:
An early return has the same semantics.
With the pointer-trick mentioned in another comment you get rid of the final `UA_free(data)`.

Ideally, the "happy case" of the control flow is linear and does not flow through nested blocks with `{ }` braces.
That can be achieved by branching out the edge cases with an early return.

As often, it is a question of taste.
Here I would prefer an early return.

Which case is the "happy case", I let you decide. 😄